Add RFC for GPUX Dialect (#323)
intel/mlir-extensions: docs/rfcs/GPUXDialect.md

#### gpux.alloc (gpux::AllocOp)

GPU memory allocation operation.

Syntax:

```
operation ::= gpux.alloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              ($gpux_stream, $dynamicSizes) ([ $symbolOperands^ ])? attr-dict : type($memref)
```

This op is an extension of upstream gpu.alloc op with one added argument for stream.

#### gpux.dealloc (gpux::DeallocOp)

GPU memory deallocation operation.

Syntax:

```
operation ::= gpux.dealloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              ($gpux_stream, $memref) attr-dict : type($memref)
```

This op is an extension of upstream gpu.dealloc op with one added argument for stream.

#### gpux.launch_func (gpux::LaunchFuncOp)

Launches a function as a GPU kernel.

Syntax:

```
operation ::= gpux.launch_func custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              $kernel
              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?

>               custom<LaunchFuncOperands>($gpux_stream, $operands, type($operands)) attr-dict

Reviewer: Why $gpux_stream is part of device function parameter? It should be part of GPUX.launch_func(), so becomes the host code interacting with runtime, not the device kernel.

Reviewer: What is the reason we don't support gpu.launch also?

Author: yes, thats a mistake. it has to be next to kernel. will change it

We dont support gpu.launch is because we assume the user to run the upstream GpuKernelOutliningPass which would translate the gpu.launch to gpu.launch_func and outline the device kernel. Most of the lower passes expect the kernel to be outlined

## Proposal

We propose to have the GPUX dialect with ops listed below:

### Operation:

#### gpux.create_device (gpux::CreateDeviceOp)

Create the GPU Device.

Syntax:

```
operation ::= gpux.create_device $device : (!gpux.DeviceType) -> !gpux.DeviceType
```

> The gpux.create_device op takes a pointer (opaque) to the underlying device and will return a sycl/l0 device handle.

Reviewer 2: ```suggestion
The gpux.create_device op takes a pointer (opaque) to an underlying device and will return a sycl/l0 device handle.
```

#### gpux.create_context (gpux::CreateContextOp)

Create the Context.

Syntax:

```
operation ::= gpux.create_context $context : (!gpux.ContextType) -> !gpux.ContextType
```

> The gpux.create_context op takes a pointer (opaque) to the underlying context and will return a sycl/l0 context handle.

Reviewer 2: Nit
```suggestion
The gpux.create_context op takes a pointer (opaque) to an underlying context and will return a sycl/l0 context handle.
```

#### gpux.create_stream (gpux::CreateStreamOp)

Create the GPU Stream.

Syntax:

```
operation ::= gpux.create_stream $device, $context: (!gpux.DeviceType, !gpux.ContextType) -> !gpux.StreamType
```

The gpux.create_stream operation will create a sycl/l0 queue with the device and context provided.

#### gpux.destroy_device (gpux::DestroyDeviceOp)

Destroy the Device.

Syntax:

```
operation ::= gpux.destroy_device $device : (!gpux.DeviceType) -> ()
```

> The gpux.destroy_device operation will destroy the above created device.

Reviewer 2: ```suggestion
The gpux.destroy_device operation will deallocate the passed in device pointer.
```

#### gpux.destroy_context (gpux::DestroyContextOp)

Destroy the Context.

Syntax:

```
operation ::= gpux.destroy_context $context : (!gpux.ContextType) -> ()
```

> The gpux.destroy_context operation will destroy the above created context.

Reviewer 2: ```suggestion
The gpux.destroy_context operation will deallocate the passed in context pointer.
```

#### gpux.destroy_stream (gpux::DestroyStreamOp)

Destroy the GPU Stream.

Syntax:

```
operation ::= gpux.destroy_stream $gpux_stream : (!gpux.StreamType) -> ()
```

> The gpux.destroy_stream() operation will destroy the above created stream. If the user wants to use their stream, we expect the user to manage the lifetime of the stream.

Reviewer 2: Same change as `destroy_device` and `destroy_context` ops.

#### gpux.alloc (gpux::AllocOp)

GPU memory allocation operation.

Syntax:

```
operation ::= gpux.alloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              ($gpux_stream, $dynamicSizes) ([ $symbolOperands^ ])? attr-dict : type($memref)
```

> This op is an extension of upstream gpu.alloc op with one added argument for stream.

Reviewer 2: ```suggestion
This op is an extension of upstream gpu.alloc op with one added argument for stream using which the memory is allocated on a specific device.
```

#### gpux.dealloc (gpux::DeallocOp)

GPU memory deallocation operation.

Syntax:

```
operation ::= gpux.dealloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              ($gpux_stream, $memref) attr-dict : type($memref)
```

> This op is an extension of upstream gpu.dealloc op with one added argument for stream.

Reviewer 2: Same feedback as alloc.

#### gpux.launch_func (gpux::LaunchFuncOp)

Launches a function as a GPU kernel.

Syntax:

```
operation ::= gpux.launch_func custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              $gpux_stream,
              $kernel,
              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?
              custom<LaunchFuncOperands>($operands, type($operands)) attr-dict
```

> This op is an extension of upstream gpu.launch_func op with one added argument for stream.

Reviewer 2: ```suggestion
This op is an extension of upstream gpu.launch_func op with one added argument for stream on which the kernel is enqueued or launched.
```

#### gpux.wait (gpux::WaitOp)

Wait for gpu ops in a particular stream to complete.

Syntax:

```
operation ::= gpu.wait custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              $gpux_stream attr-dict
```

The wait op will need to be provided with the stream to wait on. This op is an extension of upstream gpu.wait op with one added argument for stream.

#### gpux.memcpy (gpux::MemcpyOp)

GPU memcpy operation

Syntax:

```
operation ::= gpu.memcpy custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              $gpux_stream, $dst, $src : type($dst), type($src) attr-dict
```

> The gpux.memcpy operation copies the content of one memref to another. This op is an extension of upstream gpu.memcpy op with one added argument for stream.

Reviewer 2: Again, some justification is needed for why the stream arg is needed.

              $kernel
              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?
              custom<LaunchFuncOperands>($gpux_stream, $operands, type($operands)) attr-dict
```

This op is an extension of upstream gpu.launch_func op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.

#### gpux.wait (gpux::WaitOp)

Wait for gpu ops in a particular stream to complete.

Syntax:

```
operation ::= gpu.wait custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
              $gpux_stream attr-dict
```

The wait op will need to be provided with the stream to wait on. This op is an extension of upstream gpu.wait op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.

## Example Usage in case of user provided stream:

// IR before our custom transformation pass:

```

> func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {

Reviewer: This IR doesn't make sense. When user transforms to GPU dialect, user either choose to not specify stream. If user choose to user stream, it is user's responsibility to use the stream to GPU ops.

Author: >>>it is user's responsibility to use the stream to GPU ops.

how will the user pass the stream to gpu ops ? they have to use this GPUX dialect to pass the stream to GPU ops...are you saying user will have their own pass to convert to the GPUX dialect which adds stream as an input to the upstream gpu ops?

  %cst = arith.constant 2.200000e+00 : f32
  %cst_0 = arith.constant 1.100000e+00 : f32
  %cst_1 = arith.constant 0.000000e+00 : f32
  %0 = gpu.alloc  () : memref<8xf32>
  %1 = gpu.alloc  () : memref<8xf32>
  %2 = gpu.alloc  () : memref<8xf32>
  call @fillResource1DFloat(%0, %cst_0) : (memref<?xf32>, f32) -> ()
  call @fillResource1DFloat(%1, %cst) : (memref<?xf32>, f32) -> ()
  call @fillResource1DFloat(%2, %cst_1) : (memref<?xf32>, f32) -> ()
  gpu.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%0 : memref<8xf32>, %1 : memref<8xf32>, %2 : memref<8xf32>)
  %3 = memref.cast %2 : memref<8xf32> to memref<*xf32>
  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
  gpu.dealloc(%0) : memref<8xf32>
  gpu.dealloc(%1) : memref<8xf32>
  gpu.dealloc(%2) : memref<8xf32>
  return
}

```

// IR after our custom transformation pass:

```

func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {
  %c8 = arith.constant 8 : index
  %c1 = arith.constant 1 : index
  %cst = arith.constant 2.200000e+00 : f32
  %cst_0 = arith.constant 1.100000e+00 : f32
  %cst_1 = arith.constant 0.000000e+00 : f32

>   %0 = gpux.alloc (%stream) : (!gpux.StreamType) -> memref<8xf32>

Reviewer: It doesn't make sense. How the optimization pass would know that it needs tot take the stream passed by the function and assign it to this operations?  What about user pass two streams in the main function parameters? What if the operations within  program itself needs to two streams?  There should be a separate optimization pass to decide which stream to use, and that optimization pass need to decide which op to use which stream. Not just pass it as parameter, hoping another optimization pass to fix the rest.

Reviewer 3: @Jianhui-Li That's the part which do not understand. The mere existence of a stream-creation operation somewhere in the code give no hint about where to use it. Hence we suggested using regions (see #283). Can you explain how an analysis would be able to do this mapping? What's the contract for linking create_stream to the alloc?

  gpux.dealloc(%stream, %2) : (!gpux.StreamType, memref<8xf32>) -> ()
  return
}
```

Since the stream is already provided by the user in this case, only the alloc, dealloc and the launch_func operations were translated to the extended versions in the GPUX dialect.

## Example Usage in case of default stream:

// IR before our custom transformation pass:

```

func.func @main() attributes {llvm.emit_c_interface} {
  %c8 = arith.constant 8 : index
  %c1 = arith.constant 1 : index
  %cst = arith.constant 2.200000e+00 : f32
  %cst_0 = arith.constant 1.100000e+00 : f32
  %cst_1 = arith.constant 0.000000e+00 : f32
  %0 = gpu.alloc  () : memref<8xf32>
  %1 = gpu.alloc  () : memref<8xf32>
  %2 = gpu.alloc  () : memref<8xf32>
  gpu.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%0 : memref<8xf32>, %1 : memref<8xf32>, %2 : memref<8xf32>)
  %3 = memref.cast %2 : memref<8xf32> to memref<*xf32>
  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
  gpu.dealloc(%0) : memref<8xf32>
  gpu.dealloc(%1) : memref<8xf32>
  gpu.dealloc(%2) : memref<8xf32>
  return
}

```

// IR after our custom transformation pass:

```

> func.func @main() attributes {llvm.emit_c_interface} {

Reviewer: This pass makes sense to me.  This means user doesn't want to control at all.

Reviewer 3: Why not? From our perspective that's exactly what happens with compute-follows-data. There can be a default behavior, but it seems important to be able to control this.

# RFC: GPUX Dialect (Extension of upstream GPU Dialect)

Core MLIR Team

## Summary

> We propose the GPUX dialect as an extension of the upstream GPU dialect. The GPUX dialect allows us to expose stream creation/destruction ops.These ops are required by the underlying runtimes (Level zero & Sycl) for explicit stream/context/device creation.

Reviewer 4: I think you can switching the order of the statements may look more natural. You can first mention some runtimes require explicit stream context and device since that is the reason for introducing GPUX dialect. And then move onto introducing new ops (stream creation/destruction) and op changes.

The GPUX dialect will also have some of the upstream GPU dialect ops extended with an added argument for stream in those ops.

## Motivation

> Upstream MLIR has a Dialect called the [GPU Dialect](https://mlir.llvm.org/docs/Dialects/GPU/) which provides middle-level abstractions for launching GPU kernels following a programming model similar to that of CUDA or OpenCL. The upstream dialect exposes all the operation required to launch the kernel on a GPU device. But, the drawback with the upstream dialect is it does not expose Stream as an operation in the dialect. For example, if the user wants to launch the kernel on a particular stream(provided by the user), there is no way to do that today. For the upstream dialect, stream is created internally while lowering to LLVM runtime calls and it refers specifically to the CUDA stream. Therefore, the need to create a GPUX Dialect arises. The GPUX dialect is an extension of the upstream GPU dialect with following ops - CreateStreamOp, DestroyStreamOp, LaunchFuncOp, AllocOp, DeallocOp & WaitOp . The CreateStreamOp op allows for explicit Stream creation. The stream is a temporary stream and is internally creating a SYCL/L0 queue, with default context and device.

Reviewer 4: Although you have already mentioned new ops in the summary above, pointing out which ops are new and which ops are an extension of an existing GPU dialect op would help readers get a better picture.
